# <div style = "padding: 20px; border-radius: 5px; background-color: #311B92;color: white"><center>PRE-PROCESSING STAGE-I - DATASET A and B</center></DIV>

## SCOPE

- Processing Dataset A and Dataset B 
- The scope of the 1st stage of pre-processing is to check if the data of all the cars for all the 3 sections were extracted properly.
- Dataset A contains data for all the cities.
- Dataset B contains data for all cities with majority of data from cities- CHENNAI, HYDERABAD, BANGALORE, DELHI, MUMBAI

## IMPORTING LIBRARIES

In [1]:
# importing required libraries for pre-processing

import pandas as pd
import json                      # to read the json file

import os

- reading 1st batch data to very

In [2]:
# getting the current/main working directory
os.getcwd()

'C:\\Users\\pryns\\OneDrive\\Documents\\Python Works\\zz.carwale data'

## <div style = "padding: 20px; border-radius: 40px; background-color: #40E0D0"> Dataset A </div>

### CHECKING `Overview` DATA

In [10]:
# looping through the directories within the main directory and reading the car overview data

car_overview_data = []
data_len = []

file = 'car_overview.json'
path = r"C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\01.Data Extraction\carwale_dataset_A"


# returns the list of directories within the main directory
for dir in list(os.walk(path))[0][1]:
    with open(os.path.join(path,dir,file),'r') as f:
        car_data = json.loads(f.read())
    data_len.append(len(car_data))          # storing the length of each file for verification purpose
    car_overview_data.extend(car_data)
    

In [11]:
# 16 folders were checked
len(data_len)

16

In [12]:
# checking the no. of car records extracted
len(car_overview_data)

112559

In [13]:
# verifying no. of records against each file
sum(data_len)

112559

In [14]:
# converting the list of dictionaries into dataframe

df_overview_A = pd.DataFrame(car_overview_data)

In [15]:
df_overview_A.head(2)

carId profileId                                          vehicle  \
0      1  S2725575                Mercedes-Benz GLE 300d 4MATIC LWB   
1      2  S2730115  Maruti Suzuki Vitara Brezza VDi (O) [2016-2018]   

          city            state    price        color kilometers  \
0  A&N Islands  Andaman Nicobar  75 Lakh  Polar White      6,000   
1  A&N Islands  Andaman Nicobar   6 Lakh        White     55,000   

          fuelName transmissionType  ...  formattedRegistrationDate  \
0  Diesel + Diesel        Automatic  ...              Not Available   
1  Diesel + Diesel           Manual  ...              Not Available   

     fuelEconomy dealershipLogoUrl isSold oemVehicleUrl virtualPhoneNumber  \
0  Not Available                    False           NaN                NaN   
1  Not Available                    False           NaN                NaN   

  homeTestDriveSlug formattedOriginalPrice stockId additionalFuelName  
0               NaN                    NaN     NaN                NaN  
1               NaN                    NaN     NaN                NaN  

[2 rows x 61 columns]

In [17]:
df_overview_A.columns

Index(['carId', 'profileId', 'vehicle', 'city', 'state', 'price', 'color',
       'kilometers', 'fuelName', 'transmissionType', 'sellerId', 'insurance',
       'insuranceExpiry', 'interiorColor', 'lifeTimeTax', 'insuranceLink',
       'makeName', 'modelName', 'versionName', 'mainImageUrl', 'photoCount',
       'makeYear', 'makeMonth', 'cityMaskingName', 'cityName', 'cityId',
       'noOfOwners', 'registerCity', 'carAvailbaleAt', 'lastUpdatedDate',
       'totalPhotosUploaded', 'similarCarsUrl', 'stockRecommendationUrl',
       'cwBasePackageId', 'ctePackageId', 'modelId', 'versionId',
       'registrationNumber', 'tcStockId', 'regType', 'priceNumeric',
       'isCertified', 'videoUrl', 'modelMaskingName', 'rootName',
       'bodyStyleId', 'isSellCarOfferAvailable', 'allowBooking', 'emiPrice',
       'shouldShowEmiSlug', 'isHomeTestDrive', 'formattedRegistrationDate',
       'fuelEconomy', 'dealershipLogoUrl', 'isSold', 'oemVehicleUrl',
       'virtualPhoneNumber', 'homeTestDriveSlug', 

In [18]:
# checking the dataframe columns and datatypes

df_overview_A.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112559 entries, 0 to 112558
Data columns (total 61 columns):
 #   Column                     Non-Null Count   Dtype 
---  ------                     --------------   ----- 
 0   carId                      112559 non-null  int64 
 1   profileId                  112559 non-null  object
 2   vehicle                    112559 non-null  object
 3   city                       112559 non-null  object
 4   state                      112559 non-null  object
 5   price                      112559 non-null  object
 6   color                      112559 non-null  object
 7   kilometers                 112559 non-null  object
 8   fuelName                   112559 non-null  object
 9   transmissionType           112559 non-null  object
 10  sellerId                   112559 non-null  int64 
 11  insurance                  112559 non-null  object
 12  insuranceExpiry            112559 non-null  object
 13  interiorColor              112559 non-null  

In [19]:
df_overview_A.shape

(112559, 61)

- in total there are 112559 records (including duplicates)

- `profileId` are distinct car identities and will be used to identify each car
- there are 59708 unique cars from various cities (excluding the duplicates)

In [20]:
df_overview_A['profileId'].unique().shape

(59708,)

### CHECKING  `Specifications` AND `Features` DATA

- using a similar strategy as used for overview data above to consolidate the specification and features data in a separate file


In [23]:
# creating a function to consolidate specs/features into a single dataframe

def extract_combine(file_name):
    file = file_name
    
    # main path
    path = r"C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\01.Data Extraction\carwale_dataset_A"

    df_len = []  # storing the length of each file for verification purpose

    # returns the list of directories within the main directory
    # looping through the directories and reading the specifications data and combining them into a single dataframe
    
    for dir in list(os.walk(path))[0][1]:
        df_item = pd.read_csv(os.path.join(path,dir,file))
        try:
            df_items = pd.concat([df_items,df_item])
        except:
            cols = df_item.columns
            df_items = pd.DataFrame(columns = cols)
            df_items = pd.concat([df_items,df_item])
        df_len.append(df_item.shape[0])  # storing the length of each file for verification purpose

    return (df_items,df_len)     #returning the final dataframe and the list of lengths of each file

#### SPECIFICATIONS

In [24]:
# consolidating specifications
df_specs_A, df_spec_len_A = extract_combine('car_specs.csv')

In [25]:
df_specs_A.head(2)

specName specValue specUnit          spec_category carId  \
0                  Top Speed       225     Kmph  Engine & Transmission     1   
1  Acceleration (0-100 kmph)       7.2  seconds  Engine & Transmission     1   

  profileId                            vehicle         city            state  
0  S2725575  Mercedes-Benz GLE 300d 4MATIC LWB  A&N Islands  Andaman Nicobar  
1  S2725575  Mercedes-Benz GLE 300d 4MATIC LWB  A&N Islands  Andaman Nicobar

In [26]:
# verifying the no. of data files extracted - 16 - which is correct
len(df_spec_len_A)

16

In [27]:
df_specs_A.shape

(3191958, 9)

In [28]:
sum(df_spec_len_A)

3191958

In [29]:
#verifying each file records with the consilidated dataframe
sum(df_spec_len_A) == df_specs_A.shape[0]

True

#### FEATURES

In [30]:
# consolidating features
df_feats_A, df_feat_len_A = extract_combine('car_features.csv')

In [31]:
df_feats_A.head(2)

itemName itemValue feature_category carId profileId  \
0       Overspeed Warning      True           Safety     1  S2725575   
1  Lane Departure Warning     False           Safety     1  S2725575   

                             vehicle         city            state  
0  Mercedes-Benz GLE 300d 4MATIC LWB  A&N Islands  Andaman Nicobar  
1  Mercedes-Benz GLE 300d 4MATIC LWB  A&N Islands  Andaman Nicobar

In [32]:
len(df_feat_len_A)

16

In [33]:
df_feats_A.shape

(15897790, 8)

In [34]:
#verifying each file records with the consilidated dataframe
sum(df_feat_len_A) == df_feats_A.shape[0]

True

### VERIFYING DATA

- `profileId` is a unique identifier of each record/car
- this identifier is used to check if all the data in each section is available for all the cars

- finding unique `profileId` values for overview, specifications and features data first
- as this is the main data against which we will be combining other data.

In [35]:
print("unique overview records:      ",df_overview_A['profileId'].unique().shape[0])
print("unique specification records: ",df_specs_A['profileId'].unique().shape[0])
print("unique feature records:       ",df_feats_A['profileId'].unique().shape[0])

unique overview records:       59708
unique specification records:  58908
unique feature records:        58541


- some difference in records w.r.t profileId can be found between overview and specifications as well as features

In [36]:
len(set(df_overview_A['profileId']).symmetric_difference(set(df_specs_A['profileId'])))

800

In [37]:
len(set(df_overview_A['profileId']).symmetric_difference(set(df_feats_A['profileId'])))

1167

In [38]:
# BETWEEN FEATURES AND SPECIFICATIONS
len(set(df_specs_A['profileId']).symmetric_difference(set(df_feats_A['profileId'])))

443

* specification lacks 800 car records which is in overview
* features lacks 1167 car records found in overview
* lacking data from overview can be eliminated during EDA, as features data will not be used much for prediction and most of the EDA
* examining Dataset B

## <div style = "padding: 20px; border-radius: 40px; background-color: #9FE2BF"> Dataset B </div>

- using a similar strategy as used for dataset A above to consolidate the specification and features data in a separate file
- since all the files are organised in categories, its easy to use a single function for consolidation

In [40]:
# function to consolidate overview/specs/features into a single dataframe

def extract_combine(folder):
    dir = folder
    
    # main path
    path = r"C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\01.Data Extraction\carwale_dataset_B"
           
    folderpath = os.path.join(path,dir)
    
    df_len = []      # storing the length of each file for verification purpose
    
    # returns the list of directories within the main directory
    # looping through the directories and reading the specifications data and combining them into a single dataframe
    
    for file in list(os.walk(folderpath))[0][2]:
        df_item = pd.read_csv(os.path.join(folderpath,file))
        try:
            df_items = pd.concat([df_items,df_item])
        except:
            cols = df_item.columns
            df_items = pd.DataFrame(columns = cols)
            df_items = pd.concat([df_items,df_item])
        df_len.append(df_item.shape[0])    # storing the length of each file for verification purpose
        print(file,df_item.shape)
    return (df_items,df_len)     #returning the final dataframe and the list of lengths of each file

### CHECKING `Overview` DATA

In [41]:
df_overview_B, df_overview_len_B = extract_combine('overview')

car_data_overview_bangalore.csv (6639, 57)
car_data_overview_chennai.csv (3420, 57)
car_data_overview_delhi.csv (9012, 57)
car_data_overview_hyderabad.csv (4582, 57)
car_data_overview_mumbai.csv (7864, 57)


In [42]:
# verifying the no. of files checked
len(df_overview_len_B)

5

In [44]:
df_overview_B.head(2)

carId profileId                                       vehicle       city  \
0     0  D4135089                  Tata Harrier XZ Dark Edition  Bangalore   
1     1  D4170587  BMW 3 Series GT 320d Luxury Line [2014-2016]  Bangalore   

        price  color kilometers fuelName transmissionType sellerId  ...  \
0  18.95 Lakh  Black     39,420   Diesel           Manual        1  ...   
1  34.95 Lakh   Grey     38,703   Diesel        Automatic        1  ...   

  allowBooking virtualPhoneNumber emiPrice isHomeTestDrive  \
0        False       7.303409e+09   31,469           False   
1        False       7.303409e+09   58,040           False   

                                   homeTestDriveSlug  \
0  {'titlePrefix': '', 'title': 'Home Test Drive'...   
1  {'titlePrefix': '', 'title': 'Home Test Drive'...   

  formattedRegistrationDate    fuelEconomy dealershipLogoUrl isSold  \
0             Not Available  Not Available               NaN  False   
1             Not Available  Not Available               NaN  False   

  formattedOriginalPrice  
0                    NaN  
1                    NaN  

[2 rows x 57 columns]

In [47]:
df_overview_B.shape[0]

31517

In [48]:
#verifying each file records with the consilidated dataframe

df_overview_B.shape[0] == sum(df_overview_len_B)

True

### CHECKING `Specifications` DATA

In [49]:
df_specs_B, df_spec_len_B  = extract_combine('specifications')

car_data_specs_bangalore.csv (186313, 8)
car_data_specs_chennai.csv (97551, 8)
car_data_specs_delhi.csv (261004, 8)
car_data_specs_hyderabad.csv (129858, 8)
car_data_specs_mumbai.csv (219475, 8)


In [50]:
# verifying the no. of files checked
len(df_spec_len_B)

5

In [51]:
df_specs_B.head(2)

specName                                          specValue specUnit  \
0       Engine  1956 cc, 4 Cylinders Inline, 4 Valves/Cylinder...      NaN   
1  Engine Type                                      2.0 L Kryotec      NaN   

           spec_category carId profileId                       vehicle  \
0  Engine & Transmission     0  D4135089  Tata Harrier XZ Dark Edition   
1  Engine & Transmission     0  D4135089  Tata Harrier XZ Dark Edition   

        city  
0  Bangalore  
1  Bangalore

In [52]:
df_specs_B.shape[0]

894201

In [53]:
#verifying each file records with the consilidated dataframe

df_specs_B.shape[0] == sum(df_spec_len_B)

True

### CHECKING `Features` DATA

In [54]:
df_feats_B, df_feat_len_B = extract_combine('features')

car_data_features_bangalore.csv (929541, 7)
car_data_features_chennai.csv (485335, 7)
car_data_features_delhi.csv (1301337, 7)
car_data_features_hyderabad.csv (646127, 7)
car_data_features_mumbai.csv (1105963, 7)


In [55]:
# verifying the no. of files checked
len(df_feat_len_B)

5

In [56]:
df_feats_B.head(2)

itemName itemValue feature_category carId profileId  \
0       Overspeed Warning      True           Safety     0  D4135089   
1  Lane Departure Warning     False           Safety     0  D4135089   

                        vehicle       city  
0  Tata Harrier XZ Dark Edition  Bangalore  
1  Tata Harrier XZ Dark Edition  Bangalore

In [57]:
df_feats_B.shape[0]

4468303

In [58]:
#verifying each file records with the consilidated dataframe

df_feats_B.shape[0] == sum(df_feat_len_B)

True

### VERIFYING DATA

- `profileId` is a unique identifier of each record/car
- this identifier is used to check if all the data in each section is available for all the cars

- finding unique `profileId` values for overview, specifications and features data first
- as this is the main data against which we will be combining other data.

In [59]:
print("unique overview records:      ",df_overview_B['profileId'].unique().shape[0])
print("unique specification records: ",df_specs_B['profileId'].unique().shape[0])
print("unique feature records:       ",df_feats_B['profileId'].unique().shape[0])

unique overview records:       30274
unique specification records:  30474
unique feature records:        30265


- a huge difference can be found between overview and specifications as well as features

In [60]:
len(set(df_overview_B['profileId']).symmetric_difference(set(df_specs_B['profileId'])))

1076

In [61]:
len(set(df_overview_B['profileId']).symmetric_difference(set(df_feats_B['profileId'])))

1281

In [62]:
# BETWEEN FEATURES AND SPECIFICATIONS
len(set(df_specs_B['profileId']).symmetric_difference(set(df_feats_B['profileId'])))

239

* specification lacks 1076 car records which is in overview
* features lacks 1281 car records found in overview
* lacking data from overview can be eliminated during EDA, as features data will not be used much for prediction and most of the EDA

## SAVING THE FILES FROM STAGE-I PRE-PROCESSING

In [63]:
# dataset A
df_overview_A.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\overview_A.csv',index=False)
df_specs_A.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\specifications_A.csv',index=False)
df_feats_A.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\features_A.csv',index=False)

In [64]:
# dataset B
df_overview_B.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\overview_B.csv',index=False)
df_specs_B.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\specifications_B.csv',index=False)
df_feats_B.to_csv(r'C:\Users\pryns\OneDrive\Documents\Python Works\zz.carwale data\02.Pre-Processing\Pre-processed Data Stage-I\features_B.csv',index=False)